File: camera_calibrate/stereo_calibrate.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/calib3d.hpp>
#include <opencv2/core.hpp>
#include <iostream>
#include <stdio.h>

using namespace std;

const int imageWidth = 640;                             //摄像头的分辨率
const int imageHeight = 480;
const int boardWidth = 9;                               //横向的角点数目
const int boardHeight = 6;                              //纵向的角点数据
const int boardCorner = boardWidth * boardHeight;       //总的角点数据
const int frameNumber = 13;                             //相机标定时需要采用的图像帧数
const int squareSize = 25;                              //标定板黑白格子的大小 单位mm
const cv::Size boardSize = cv::Size(boardWidth, boardHeight);   //标定板的总内角点
cv::Size imageSize = cv::Size(imageWidth, imageHeight);

cv::Mat R, T, E, F;                                                  //R 旋转矢量 T平移矢量 E本征矩阵 F基础矩阵
vector<cv::Mat> rvecs;                                        //旋转向量
vector<cv::Mat> tvecs;                                        //平移向量
vector<vector<cv::Point2f>> imagePointL;                    //左边摄像机所有照片角点的坐标集合
vector<vector<cv::Point2f>> imagePointR;                    //右边摄像机所有照片角点的坐标集合
vector<vector<cv::Point3f>> objRealPoint;                   //各副图像的角点的实际物理坐标集合

vector<cv::Point2f> cornerL;                              //左边摄像机某一照片角点坐标集合

vector<cv::Point2f> cornerR;                              //右边摄像机某一照片角点坐标集合

cv::Mat rgbImageL, grayImageL;
cv::Mat rgbImageR, grayImageR;

cv::Mat Rl, Rr, Pl, Pr, Q;                                  //校正旋转矩阵R，投影矩阵P 重投影矩阵Q (下面有具体的含义解释）
cv::Mat mapLx, mapLy, mapRx, mapRy;                         //映射表
cv::Rect validROIL, validROIR;                              //图像校正之后，会对图像进行裁剪，这里的validROI就是指裁剪之后的区域

/*
事先标定好的左相机的内参矩阵
fx 0 cx
0 fy cy
0 0  1
*/
cv::Mat cameraMatrixL = (cv::Mat_<double>(3, 3) << 536.0734301865696, 0, 342.37038790606886,
                         0, 536.0163447473561, 235.53685635127164,
                         0, 0, 1);                                                                           //这时候就需要你把左右相机单目标定的参数给写上
//获得的畸变参数
cv::Mat distCoeffL = (cv::Mat_<double>(5, 1) << -0.2650900636135801, -0.04674390481455304, 0.0018330102869313922, -0.0003147146607548382, 0.2523158614581943);
/*
事先标定好的右相机的内参矩阵
fx 0 cx
0 fy cy
0 0  1
*/
cv::Mat cameraMatrixR = (cv::Mat_<double>(3, 3) << 542.3547430679187, 0, 328.3241889687017,
                         0, 541.6149960760252, 246.94729219873494,
                         0, 0, 1);
cv::Mat distCoeffR = (cv::Mat_<double>(5, 1) << -0.28054308254579113, 0.1043237314713875, -0.0005582136249751141, 0.0013035577012644342, -0.02372163112169363);


/*计算标定板上模块的实际物理坐标*/
void calRealPoint(std::vector<std::vector<cv::Point3f>>& obj, int boardwidth, int boardheight, int imgNumber, int squaresize)
{
    std::vector<cv::Point3f> imgpoint;
    for (int rowIndex = 0; rowIndex < boardheight; rowIndex++)
    {
        for (int colIndex = 0; colIndex < boardwidth; colIndex++)
        {
            imgpoint.push_back(cv::Point3f(rowIndex * squaresize, colIndex * squaresize, 0));
        }
    }
    for (int imgIndex = 0; imgIndex < imgNumber; imgIndex++)
    {
        obj.push_back(imgpoint);
    }
}

void outputCameraParam(void)
{
    /*保存数据*/
    /*输出数据*/
    cv::FileStorage fs("intrinsics.yml", cv::FileStorage::WRITE);  //文件存储器的初始化
            if (fs.isOpened())
    {
        fs << "cameraMatrixL" << cameraMatrixL << "cameraDistcoeffL" << distCoeffL << "cameraMatrixR" << cameraMatrixR << "cameraDistcoeffR" << distCoeffR;
        fs.release();
        std::cout << "cameraMatrixL=:" << cameraMatrixL << endl << "cameraDistcoeffL=:" << distCoeffL << endl << "cameraMatrixR=:" << cameraMatrixR << endl << "cameraDistcoeffR=:" << distCoeffR << endl;
    }
    else
    {
        std::cout << "Error: can not save the intrinsics!!!!!" << endl;
    }


    fs.open("extrinsics.yml", cv::FileStorage::WRITE);
    if (fs.isOpened())
    {
        fs << "R" << R << "T" << T << "Rl" << Rl << "Rr" << Rr << "Pl" << Pl << "Pr" << Pr << "Q" << Q;
        std::cout << "R=" << R << endl << "T=" << T << endl << "Rl=" << Rl << endl << "Rr=" << Rr << endl << "Pl=" << Pl << endl << "Pr=" << Pr << endl << "Q=" << Q << endl;
        fs.release();
    }
    else
        std::cout << "Error: can not save the extrinsic parameters\n";
}

int main()
{
    cv::Mat img;
    int goodFrameCount = 0;
    while (goodFrameCount < frameNumber)
    {
        char filename[100];
        /*读取左边的图像*/
        sprintf(filename, "/Users/han/Company/opencv-3.4.1/samples/data/calibrare_img/left/left%02d.jpg", goodFrameCount + 1);

        rgbImageL = cv::imread(filename, CV_LOAD_IMAGE_COLOR);

        cv::cvtColor(rgbImageL, grayImageL, CV_BGR2GRAY);

        /*读取右边的图像*/
        sprintf(filename, "/Users/han/Company/opencv-3.4.1/samples/data/calibrare_img/right/right%02d.jpg", goodFrameCount + 1);
        rgbImageR = cv::imread(filename, CV_LOAD_IMAGE_COLOR);
        cv::cvtColor(rgbImageR, grayImageR, CV_BGR2GRAY);

        bool isFindL, isFindR;

        isFindL = cv::findChessboardCorners(rgbImageL, boardSize, cornerL);
        isFindR = cv::findChessboardCorners(rgbImageR, boardSize, cornerR);
        if (isFindL == true && isFindR == true)  //如果两幅图像都找到了所有的角点 则说明这两幅图像是可行的
        {
            cv::cornerSubPix(grayImageL, cornerL, cv::Size(5, 5), cv::Size(-1, -1), cv::TermCriteria(CV_TERMCRIT_EPS | CV_TERMCRIT_ITER, 20, 0.1));
            cv::drawChessboardCorners(rgbImageL, boardSize, cornerL, isFindL);
            cv::imshow("chessboardL", rgbImageL);
            imagePointL.push_back(cornerL);

            cv::cornerSubPix(grayImageR, cornerR, cv::Size(5, 5), cv::Size(-1, -1), cv::TermCriteria(CV_TERMCRIT_EPS | CV_TERMCRIT_ITER, 20, 0.1));
            cv::drawChessboardCorners(rgbImageR, boardSize, cornerR, isFindR);
            cv::imshow("chessboardR", rgbImageR);
            imagePointR.push_back(cornerR);

            goodFrameCount++;
            std::cout << "The image" << goodFrameCount << " is good" << std::endl;
        }
            else
        {
            std::cout << "The image is bad please try again" << std::endl;
        }

        if (cv::waitKey(10) == 'q')
        {
            break;
        }
    }

    /**
        计算实际的校正点的三维坐标
        根据实际标定格子的大小来设置
    */
    calRealPoint(objRealPoint, boardWidth, boardHeight, frameNumber, squareSize);
    std::cout << "cal real successful" << std::endl;

    /*
        标定摄像头
        由于左右摄像机分别都经过了单目标定
        所以在此处选择flag = CALIB_USE_INTRINSIC_GUESS
        */
    // 返回重投影误差
    double rms = cv::stereoCalibrate(objRealPoint, imagePointL, imagePointR,
                                     cameraMatrixL, distCoeffL,
                                     cameraMatrixR, distCoeffR,
                                     cv::Size(imageWidth, imageHeight),R, T,E, F, cv::CALIB_USE_INTRINSIC_GUESS,
                                     cv::TermCriteria(cv::TermCriteria::COUNT + cv::TermCriteria::EPS, 100, 1e-5)); //需要注意，应该是版本的原因，该函数最                                                                                                                            后两个参数，我是调换过来后才显示不出错的

    std::cout << "Stereo Calibration done with RMS error = " << rms << std::endl;

    /*
        立体校正的时候需要两幅图像共面并且行对准 以使得立体匹配更加的可靠
        使得两幅图像共面的方法就是把两个摄像头的图像投影到一个公共成像面上，这样每幅图像从本图像平面投影到公共图像平面都需要一个旋转矩阵R
        stereoRectify 这个函数计算的就是从图像平面投影到公共成像平面的旋转矩阵Rl,Rr。 Rl,Rr即为左右相机平面行对准的校正旋转矩阵。
        左相机经过Rl旋转，右相机经过Rr旋转之后，两幅图像就已经共面并且行对准了。
        其中Pl,Pr为两个相机的投影矩阵，其作用是将3D点的坐标转换到图像的2D点的坐标:P*[X Y Z 1]' =[x y w]
        Q矩阵为重投影矩阵，即矩阵Q可以把2维平面(图像平面)上的点投影到3维空间的点:Q*[x y d 1] = [X Y Z W]。其中d为左右两幅图像的时差
        */
    //对标定过的图像进行校正
    cv:: stereoRectify(cameraMatrixL, distCoeffL, cameraMatrixR, distCoeffR, imageSize, R, T, Rl, Rr, Pl, Pr, Q,
                       cv::CALIB_ZERO_DISPARITY, -1, imageSize, &validROIL, &validROIR);
    /*
                                                                根据stereoRectify 计算出来的R 和 P 来计算图像的映射表 mapx,mapy
                                                                mapx,mapy这两个映射表接下来可以给remap()函数调用，来校正图像，使得两幅图像共面并且行对准
                                                                ininUndistortRectifyMap()的参数newCameraMatrix就是校正后的摄像机矩阵。在openCV里面，校正后的计算机矩阵Mrect是跟投影矩阵P一起返回的。
                                                                所以我们在这里传入投影矩阵P，此函数可以从投影矩阵P中读出校正后的摄像机矩阵
                                                                */
    //摄像机校正映射
    cv::initUndistortRectifyMap(cameraMatrixL, distCoeffL, Rl, Pr, imageSize, CV_32FC1, mapLx, mapLy);
    cv::initUndistortRectifyMap(cameraMatrixR, distCoeffR, Rr, Pr, imageSize, CV_32FC1, mapRx, mapRy);

    cv::Mat rectifyImageL, rectifyImageR;
    cv::cvtColor(grayImageL, rectifyImageL, CV_GRAY2BGR);
    cv::cvtColor(grayImageR, rectifyImageR, CV_GRAY2BGR);

    cv::imshow("Rectify Before", rectifyImageL);
    std::cout << "按Q1退出 ..." << std::endl;

    /*
        经过remap之后，左右相机的图像已经共面并且行对准了
        */
    cv::Mat rectifyImageL2, rectifyImageR2;
    cv::remap(rectifyImageL, rectifyImageL2, mapLx, mapLy, cv::INTER_LINEAR);
    cv::remap(rectifyImageR, rectifyImageR2, mapRx, mapRy, cv::INTER_LINEAR);
    std::cout << "按Q2退出 ..." << std::endl;

    cv::imshow("rectifyImageL", rectifyImageL2);
    cv::imshow("rectifyImageR", rectifyImageR2);

//    /*保存并输出数据*/
//    outputCameraParam();

    /*
        把校正结果显示出来
        把左右两幅图像显示到同一个画面上
        这里只显示了最后一副图像的校正结果。并没有把所有的图像都显示出来
        */
    cv::Mat canvas;
    double sf;
    int w, h;
    sf = 600. / MAX(imageSize.width, imageSize.height);
    w = cvRound(imageSize.width * sf);
    h = cvRound(imageSize.height * sf);

    std::cout<<"w:"<<w<<"---"<<"h:"<<h<<std::endl ;
    canvas.create(h, w * 2, CV_8UC3);

    /*左图像画到画布上*/
    cv::Mat canvasPart = canvas(cv::Rect(w * 0, 0, w, h));                                //得到画布的一部分
    cv::resize(rectifyImageL2, canvasPart, canvasPart.size(), 0, 0, cv::INTER_AREA);        //把图像缩放到跟canvasPart一样大小
    cv::Rect vroiL(cvRound(validROIL.x*sf), cvRound(validROIL.y*sf),                //获得被截取的区域
                           cvRound(validROIL.width*sf), cvRound(validROIL.height*sf));
    cv::rectangle(canvasPart, vroiL, cv::Scalar(0, 0, 255), 3, 8);                      //画上一个矩形

    std::cout << "Painted ImageL" << std::endl;

    /*右图像画到画布上*/
    canvasPart = canvas(cv::Rect(w, 0, w, h));                                      //获得画布的另一部分
    cv::resize(rectifyImageR2, canvasPart, canvasPart.size(), 0, 0, cv::INTER_LINEAR);
    cv::Rect vroiR(cvRound(validROIR.x * sf), cvRound(validROIR.y*sf),
                   cvRound(validROIR.width * sf), cvRound(validROIR.height * sf));
    cv::rectangle(canvasPart, vroiR, cv::Scalar(0, 255, 0), 3, 8);

    std::cout << "Painted ImageR" << std::endl;

    /*画上对应的线条*/
    for (int i = 0; i < canvas.rows; i += 16)
        cv::line(canvas, cv::Point(0, i), cv::Point(canvas.cols, i), cv::Scalar(0, 255, 0), 1, 8);

    cv::imshow("rectified", canvas);

    std::cout << "wait key" << std::endl;
    cv::waitKey(0);
    return 0;
}

File: camera_calibrate/stereo_calibrate.py
#coding:utf-8

import cv2
import glob
import os
import numpy as np
import math
import json
import random



boardWidth = 6
boardHeight = 5
boardCorner = boardWidth * boardHeight
squareSize = 55
boardSize = (boardWidth , boardHeight)

objpL = np.zeros((boardHeight * boardWidth , 3) , dtype=np.float32)


class Stereo_Calibrate:
	def __init__(self , left_path , right_path , imageW = 640, imageH = 480 , boardW = 6 , boardH = 5 , squareS = 55, frameN = 20):
		self.left_path = left_path
		self.right_path = right_path
		self.imageWidth = imageW
		self.imageHeight = imageH
		self.boardWidth = boardW
		self.boardHeight = boardH
		self.squareSize = squareS
		self.frameNumber = frameN
		self.left_maps = None
		self.right_maps = None
		# 第一个相机和第二个相机之间的旋转矩阵
		self.R = None
		# 平移矩阵
		self.T = None
		# 两个相机的投影矩阵
		self.Pl = None
		self.Pr = None
		# 基础矩阵和本质矩阵
		self.F = None
		self.E = None

		self.R1 = None
		self.R2 = None

		self.cameraMatrixL = None
		self.cameraMatrixR = None
		self.cameraDistCoeffL = None
		self.cameraDistCoeffR = None

		self.imagePointL = []
		self.imagePointR = []
		self.objRealPoint = []

		self.calib_info = {}

	def readCameraParam(self , calib_params):
		load_params = {}
		with open(calib_params, 'r') as f:
			load_params = json.load(f)

		mtx = load_params["mtx"]
		dist = load_params["dist"]

		return np.squeeze(np.array([mtx])), np.squeeze(np.array([dist]))

	# 计算标定板上真实的物理坐标
	def calRealPoint(self):
		imgpoint = np.zeros((self.boardWidth * self.boardHeight, 1, 3), dtype=np.float32)
		for rowIndex in range(self.boardHeight):
			for colIndex in range(self.boardWidth):
				imgpoint[rowIndex * self.boardWidth + colIndex, 0, 0] = rowIndex * self.squareSize
				imgpoint[rowIndex * self.boardWidth + colIndex, 0, 1] = colIndex * self.squareSize
				imgpoint[rowIndex * self.boardWidth + colIndex, 0, 2] = 0

		objs = []
		for imgIndex in range(self.frameNumber):
			objs.append(imgpoint)
		return objs

	def createMap(self):

		print('cameraMatrixL:' , self.cameraMatrixL)
		print('cameraDistCoeffL:' , self.cameraDistCoeffL)
		print('R1:' , self.R1)
		print('Pl:' , self.Pl)

		self.left_maps = cv2.initUndistortRectifyMap(self.cameraMatrixL , self.cameraDistCoeffL , self.R1 , self.Pl ,
													 (self.imageWidth , self.imageHeight) , cv2.CV_16SC2)

		self.right_maps = cv2.initUndistortRectifyMap(self.cameraMatrixR, self.cameraDistCoeffR, self.R2, self.Pr,
													 (self.imageWidth, self.imageHeight), cv2.CV_16SC2)

	def stereo_rectify(self , left_img_path , right_img_path):
		left_img = cv2.imread(left_img_path)
		right_img = cv2.imread(right_img_path)

		if not self.left_maps or not self.right_maps:
			self.createMap()

		rectifyImageL = cv2.remap(left_img, self.left_maps[0], self.left_maps[1], cv2.INTER_LINEAR)
		rectifyImageR = cv2.remap(right_img, self.right_maps[0], self.right_maps[1], cv2.INTER_LINEAR)

		(height, width) = np.shape(left_img)[:2]

		rectifyBefore = np.zeros((height, 2 * width, 3), dtype=np.uint8)
		rectifyBefore[:, 0:width, :] = left_img
		rectifyBefore[:, width:, :] = right_img

		np.random.seed(0)
		for i in range(0, self.imageHeight, 32):
			cv2.line(rectifyBefore, (0, i), (2 * width, i),
			         (random.randint(0, 256), random.randint(0, 256), random.randint(0, 256)), 1, 8)

		# cv2.imshow("rectifyBefore", rectifyBefore)

		sf = 600.0 / max(width, height)
		w = int(math.floor(width * sf + 0.5))
		h = int(math.floor(height * sf + 0.5))

		canvas = np.zeros((h, w * 2, 3), dtype=np.uint8)
		# 左图画到画布上
		canvasPart = canvas[:, 0:w, :]
		canvasPart = cv2.resize(rectifyImageL, (w, h), 0, 0, cv2.INTER_AREA)

		canvas[:, 0:w, :] = canvasPart

		# 右图画到画布上
		canvasPart = canvas[:, w:, :]
		canvasPart = cv2.resize(rectifyImageR, (w, h), 0, 0, cv2.INTER_AREA)
		canvas[:, w:, :] = canvasPart

		# 画上对应的线条
		for i in range(0, canvas.shape[0], 32):
			cv2.line(canvas, (0, i), (2 * w, i),
			         (random.randint(0, 256), random.randint(0, 256), random.randint(0, 256)), 1, 8)
		# cv2.imshow("rectofied", canvas)
		# cv2.waitKey(0)

		return rectifyBefore , canvas


	def calibrate(self):
		'''
		:return: 重投影误差
		'''
		if not os.path.exists(left_path) or not os.path.exists(right_path):
			print('存储图片路径不存在，请检查！')
			return None
		if not os.path.exists(os.path.join(left_path , "cameraparam_left.json")) or not os.path.exists(os.path.join(right_path , "cameraparam_right.json")):
			print('未进行相机标定，请对单个相机进行标定！')
			return None

		cameraMatrixL, distCoeffL = self.readCameraParam(os.path.join(self.left_path, "cameraparam_left.json"))
		cameraMatrixR, distCoeffR = self.readCameraParam(os.path.join(self.right_path, "cameraparam_right.json"))

		goodFrameCount = 0
		while goodFrameCount < self.frameNumber:
			leftName = os.path.join(self.left_path, "left-%d.jpg" % (goodFrameCount))
			print(leftName)

			rgbImageL = cv2.imread(leftName)

			grayImageL = cv2.cvtColor(rgbImageL, cv2.COLOR_BGR2GRAY)

			rightName = leftName.replace("left", "right")
			rgbImageR = cv2.imread(rightName)
			grayImageR = cv2.cvtColor(rgbImageR, cv2.COLOR_BGR2GRAY)

			find_chessboard_flags = cv2.CALIB_CB_ADAPTIVE_THRESH | cv2.CALIB_CB_NORMALIZE_IMAGE | cv2.CALIB_CB_FAST_CHECK
			isFindL, cornersL = cv2.findChessboardCorners(rgbImageL, boardSize, flags=find_chessboard_flags)
			isFindR, cornersR = cv2.findChessboardCorners(rgbImageR, boardSize, flags=find_chessboard_flags)

			if isFindL and isFindR:
				criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 30, 0.1)
				cv2.cornerSubPix(grayImageL, cornersL, (5, 5), (-1, -1), criteria)
				cv2.drawChessboardCorners(rgbImageL, boardSize, cornersL, isFindL)
				cv2.imshow("chessboardL", rgbImageL)
				self.imagePointL.append(cornersL)

				cv2.cornerSubPix(grayImageR, cornersR, (5, 5), (-1, -1), criteria)
				cv2.drawChessboardCorners(rgbImageR, boardSize, cornersR, isFindR)
				cv2.imshow("chessboardR", rgbImageR)
				self.imagePointR.append(cornersR)

				goodFrameCount += 1
				print("The image ", goodFrameCount, " is good")
			else:
				print('The image is bad ,pleaes try again')

			if cv2.waitKey(40) == ord('q'):
				break
		self.objRealPoint = self.calRealPoint()

		R = None
		T = None
		E = None
		F = None

		# stereocalib_retval, cameraMatrixL, distCoeffL, cameraMatrixR, distCoeffR, self.R, self.T, self.E, self.F = cv2.stereoCalibrate(
		# 	objRealPoint, imagePointL, imagePointR, cameraMatrixL, distCoeffL, cameraMatrixR, distCoeffR,
		# 	(self.imageWidth, self.imageHeight), flags=cv2.CALIB_USE_INTRINSIC_GUESS,
		# 	criteria=(cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 100, 1e-5))

		stereocalib_retval, cameraMatrixL, distCoeffL, cameraMatrixR, distCoeffR, self.R, self.T, self.E, self.F = cv2.stereoCalibrate(
			self.objRealPoint, self.imagePointL, self.imagePointR, cameraMatrixL, distCoeffL, cameraMatrixR, distCoeffR,
			(self.imageWidth, self.imageHeight), flags=cv2.CALIB_FIX_INTRINSIC,
			criteria=(cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 100, 1e-5))

		# 0 = full crop , 1 = no crop
		rectify_scale = 0
		# Pl 相机1矫正后的投影矩阵 3 * 4
		self.R1, self.R2, self.Pl, self.Pr, Q, roi1, roi2 = cv2.stereoRectify(cameraMatrixL, distCoeffL, cameraMatrixR, distCoeffR,
		                                                  (self.imageWidth, self.imageHeight), self.R, self.T, alpha=rectify_scale)

		# self.left_maps = cv2.initUndistortRectifyMap(cameraMatrixL, distCoeffL, self.R1, self.Pl, (self.imageWidth, self.imageHeight),
		#                                         cv2.CV_16SC2)
		# self.right_maps = cv2.initUndistortRectifyMap(cameraMatrixR, distCoeffR, self.R2, self.Pr, (self.imageWidth, self.imageHeight),
		#                                          cv2.CV_16SC2)

		self.cameraMatrixL = cameraMatrixL
		self.cameraMatrixR = cameraMatrixR
		self.cameraDistCoeffL = distCoeffL
		self.cameraDistCoeffR = distCoeffR

		self.createMap()



		return stereocalib_retval

	def getCalibError(self):

		err = 0
		npoints = 0
		for i in xrange(len(self.objRealPoint)):
			npt = len(self.objRealPoint[i])

			self.imagePointL[i] = cv2.undistortPoints(self.imagePointL[i] , self.cameraMatrixL , self.cameraDistCoeffL )
			self.imagePointR[i] = cv2.undistortPoints(self.imagePointR[i] , self.cameraMatrixR , self.cameraDistCoeffR )

			line_L = cv2.computeCorrespondEpilines(self.imagePointL[i] , 2 , self.F)
			line_R = cv2.computeCorrespondEpilines(self.imagePointR[i] , 1 , self.F)

			line_L = line_L.reshape(-1 , 3)
			line_R = line_R.reshape(-1 , 3)


			# # 计算误差
			for j in xrange(npt):
				err_ij = math.fabs(self.imagePointL[i][j][0][0] * line_R[j][0] + self.imagePointL[i][j][0][1] * line_R[j][1] + line_R[j][2]) + math.fabs(self.imagePointR[i][j][0][0] * line_L[j][0] + self.imagePointR[i][j][0][1] * line_L[j][1] + line_L[j][2])
				err += err_ij

			npoints += err
		# average reprojection err
		return err/npoints


	# 校正后的投影矩阵
	def getProjectRectifyMatrix(self):
		return self.Pl , self.Pr


	def getNewIntristicMatrix(self):
		return self.cameraMatrixL , self.cameraDistCoeffL , self.cameraMatrixR , self.cameraDistCoeffR

	def getRotaionTranslation(self):
		return self.R , self.T

	def getFoundationMatrix(self):
		return self.F

	def getEssentialMatrix(self):
		return self.E

	def getTranformWMatrix(self):
		'''
		计算相机1到相机2的变换矩阵
		Ar :右相机内参 4 * 4
		M  :立体标定的旋转和平移矩阵 4 * 4 [R|T]
		Al_ :左相机的内参的逆 4 * 4
		W = Ar * M * Al_
		:return: 4 x 4
		'''

		# W_matrix = np.identity(4 , dtype=np.float32)
		Ar = np.identity(4 , dtype=np.float32)
		Ar[:3 , :3] = self.cameraMatrixL
		Al = np.identity(4 , dtype=np.float32)
		Al[:3 , :3] = self.cameraMatrixL

		M = np.identity(4 , dtype=np.float32)
		M[:3,:3] = self.R
		M[:3,3:] = self.T

		print('M shape:' ,np.shape(M))

		W = np.dot(Ar , M)

		return np.dot(W , np.mat(Al).I.getA())

	def saveCalibResult(self , savefile):
		self.calib_info["cameraMatrixL"] = self.cameraMatrixL.tolist()
		self.calib_info["cameraMatrixR"] = self.cameraMatrixR.tolist()
		self.calib_info["rotation"] = self.R.tolist()
		self.calib_info["translation"] = self.T.tolist()
		self.calib_info["W"] = self.getTranformWMatrix().tolist()
		self.calib_info["F"] = self.F.tolist()
		self.calib_info["E"] = self.E.tolist()
		self.calib_info["distCoeffL"] = self.cameraDistCoeffL.tolist()
		self.calib_info["distCoeffR"] = self.cameraDistCoeffR.tolist()
		self.calib_info["R1"] = self.R1.tolist()
		self.calib_info["R2"] = self.R2.tolist()
		self.calib_info["Pl"] = self.Pl.tolist()
		self.calib_info["Pr"] = self.Pr.tolist()


		with open(savefile , "w") as f:
			json.dump(self.calib_info , f , indent=4)

	def loadCalibResuls(self , calibResult):
		camera_info = {}
		with open(calibResult , "rb") as f:
			camera_info = json.load(f)
		self.cameraMatrixL = np.array(camera_info["cameraMatrixL"])
		self.cameraMatrixR = np.array(camera_info["cameraMatrixR"])
		self.R = np.array(camera_info["rotation"])
		self.T = np.array(camera_info["translation"])
		self.cameraDistCoeffL = np.array(camera_info["distCoeffL"])
		self.cameraDistCoeffR =  np.array(camera_info["distCoeffR"])
		self.R1 = np.array(camera_info["R1"])
		self.R2 = np.array(camera_info["R2"])
		self.Pl = np.array(camera_info["Pl"])
		self.Pr = np.array(camera_info["Pr"])






if __name__ == '__main__':
	print(cv2.__version__)
	left_path = "left"
	right_path = "right"

	stereo_calib = Stereo_Calibrate(left_path , right_path , frameN=10)
	err = stereo_calib.calibrate()
	print('err:' , err)
	stereo_calib.saveCalibResult("stereo_calib_result.json")


	# ---------------------------------------
	# 立体校正
	left_img_path = "left/left-10.jpg"
	right_img_path = left_img_path.replace("left" , "right")

	rectify_before , rectified = stereo_calib.stereo_rectify(left_img_path , right_img_path)

	MatrixL , distcoeffL , MatrixR , distcoeffR = stereo_calib.getNewIntristicMatrix()
	print('-----------------------------------------------')
	print('左相机内参：')
	print(MatrixL)
	print('左相机畸变系数:')
	print(distcoeffL)
	print('-----------------------------------------------')
	print('右相机内参：')
	print(MatrixR)
	print('右相机畸变系数:')
	print(distcoeffR)
	print('-----------------------------------------------')
	R , T = stereo_calib.getRotaionTranslation()
	print('两相机旋转矩阵：')
	print(R)
	print('平移矩阵:')
	print(T)
	print('-----------------------------------------------')

	res_err = stereo_calib.getCalibError()
	print('reprojection err:' , res_err)


	cv2.imshow("before" , rectify_before)
	cv2.imshow("rectified" , rectified)
	cv2.imwrite("before.jpg" , rectify_before)
	cv2.imwrite("rectified.jpg" , rectified)
	cv2.waitKey(0)





